Added TouchBar support for MacBook Pro

sacredbanana opened the pull request:
This adds TouchBar support for MacBook Pros that have one.

It adds buttons for Compile, Toggle GUI, Textures and Quit.

Fully tested, and doesn't break other platforms.

See the image below to see what it looks like.

[](url)https://imgur.com/gallery/Cpsv4KF

Gargaj commented:
I don't think this warrants its own namespace; I'd much rather have this stashed away as an additional connection that sends keyboard commands into the renderer's message pump.

sacredbanana commented:
That sounds like a more complex solution than a simple Namespace which is only accessible from Mac devices anyway doesn't it?

sacredbanana commented:
I'm happy to change things but I'm not sure how to implement it in any other way.

Gargaj commented:
I don't think it'd be more complex and it'd keep the platform specific pollution out of main.cpp as well.

sacredbanana commented:
main.cpp would have to still execute it anyway. The user wouldn't want to launch a separate app just for this. Also main.cpp already contains apple specific #ifdeffed code :)

Gargaj commented:
It does, but it shouldn't; it's been on my list to remove all the #ifdef's from main.

Gargaj commented:
I'm addressing said issue in #69

sacredbanana commented:
Nice job. I am recoding the feature now. I will make use of the misc.cpp as per the changes you made in the `unifdef` branch. So far I have stripped away all the logic and variables it used and it instead sends keyboard commands. I was pleasantly surprised how easy apple makes sending keyboard commands in code. When you have finalised `unifdef`and merged it into `master`I'll update my PR.

Gargaj commented:
It's merged now.

sacredbanana commented:
Without using namespaces, how would you suggest to go about the call to the showTouchBar method? This method is in TouchBar.mm and is the bridge from C++ to the rest of the Objective-C code (TouchBar needs Objective-C or Swift). It needs to be called after the window has been created because the TouchBar needs to have a window pointer to attach to. I'm not sure how you feel about #ifdef within Render.cpp but I currently have an #ifdef __APPLE__ right after the window was created to call the showTouchBar method passing in the window pointer.

One idea is to create a processWindow(GLFWwindow *window) method inside Misc and any platform  can do whatever they want in there or nothing at all, but in Mac's case,  it will create the TouchBar

sacredbanana commented:
I'll go ahead with that idea and push so you can see the code to see if you like it

Gargaj commented on src/platform_glfw/Renderer.cpp:
```diff
@@ -244,6 +249,10 @@ namespace Renderer
     }
 
+#ifdef __APPLE__
+      Misc::ProcessWindow(mWindow);
```
Don't see why this is in Misc; you could just have it as an #ifdef APPLE function inside Renderer.

sacredbanana replied:
I thought Misc would be an appropriate place to put platform specific functionality. I need to call the ShowTouchBar method which needs to remain in the .mm file because .mm informs the compiler that it is an Objective-C++ file. It won't compile if I just bring that whole function out of that file

Gargaj replied:
Renderer can be platform-specific (indeed if you look at it, it has a few W32-specific pieces, although limited); it just happens to be luck that most of the OpenGL backend is the same across the board.

Putting ShowTouchBar in Renderer::Open sounds perfectly reasonable to me.

sacredbanana replied:
Well as I said ShowTouchBar has to live in TouchBar.mm or it won't compile because the compiler will treat render.cpp as a C++ file but ShowTouchBar is Objective-C++

Gargaj replied:
Yes but you can call it from the renderer - Misc::ProcessWindow is entirely unnecessary.

sacredbanana commented:
Bad commit. Disregard it. Fixing now

Gargaj commented on src/main.cpp:
```diff
@@ -11,6 +11,7 @@
 #include "UniConversion.h"
 #include "jsonxx.h"
 #include "Capture.h"
+#include "TouchBar.h"
```
Is this still needed?

sacredbanana replied:
You're right, I'll remove it.

sacredbanana commented:
Also build failed. Will work on it

sacredbanana commented:
Hehe all done :)

Gargaj commented on src/Misc.h:
```diff
@@ -1,4 +1,5 @@
 #include <string>
+#include <GLFW/glfw3.h>
```
Why?

Gargaj commented on src/main.cpp:
```diff
@@ -278,7 +278,6 @@ int main(int argc, char *argv[])
   static float fftDataSmoothed[FFT_SIZE];
   memset(fftDataSmoothed, 0, sizeof(float) * FFT_SIZE);
 
-
```
Why?

Gargaj commented on src/platform_glfw/Renderer.cpp:
```diff
@@ -244,6 +249,10 @@ namespace Renderer
     }
 
+#ifdef __APPLE__
+      ShowTouchBar(mWindow);
```
Indentation

Gargaj commented on src/platform_osx/Misc.cpp:
```diff
@@ -8,6 +8,7 @@
 
 #include <sys/param.h> // For MAXPATHLEN
 #include "CoreFoundation/CoreFoundation.h"
+#include "../TouchBar.h"
```
Not needed

Gargaj commented on src/platform_x11/Misc.cpp:
```diff
@@ -16,17 +16,17 @@ void Misc::PlatformShutdown()
 void Misc::InitKeymaps()
 {
-	return;
+    return;
```
Why

sacredbanana replied:
Damn Xcode. I noticed this too just then. I'm checking out the original files from master then recommitting

Gargaj commented on src/platform_glfw/Renderer.cpp:
```diff
@@ -23,6 +23,11 @@
 #include "Scintilla.h"
 
+#ifdef __APPLE__
+#include "Misc.h"
```
This is no longer necessary.

Gargaj replied:
Also remove the .ds_store file but I cant add a comment on that.

Gargaj commented on src/ShaderEditor.h:
```diff
@@ -1,3 +1,5 @@
+#ifndef SHADER_EDITOR_H
+#define SHADER_EDITOR_H
```
Is this needed?

Gargaj commented on src/Renderer.h:
```diff
@@ -1,3 +1,5 @@
+#ifndef RENDERER_H
+#define RENDERER_H
```
Is this needed?

Gargaj replied:
That's because TouchBar..h includes Renderer.h and ShaderEditor.h even though it needs neither.

sacredbanana replied:
Good point. 
Fixing now.

Gargaj commented:
@alkama can you give this a quick test?

sacredbanana commented:
Thank you for your help. This PR is looking much cleaner

alkama commented:
I haven't tried it yet (still have to find out how I can checkout that PR locally) but two (maybe stupid) questions arise.
Since the TouchBar API only arrived at 10.12.2, it means from now on, only macs having this version can compile from source. So maybe a toggle to ignore completely in CMake would be a good idea.

I don't know at all what happens to built binaries running on older mac either. Nor for macs that don't have a TouchBar (if it will silently gets ignored or crash).

sacredbanana commented:
I actually did have a `@available(macOS 10.12.2, *)` check but the Travis-CI build failed because API checks in Objective-C requires Xcode 9 but Travis is using Xcode 8. I went to their website and it's possible to switch to their Xcode 9.3 image. Could this please be done?

alkama commented:
Could be done with cmake injecting a define, and just wrapping all the thing in usual preprocessor thing. Thus it wouldnt depend on a special IDE/Compiler/whatever, and leave the choice to the user to include touchbar or not :)
Thought if it breaks at runtime for <10.12 systems, then releasing a binary would be made complicated indeed.

sacredbanana commented:
Xcode 9 requires macOS 10.12 which is still pretty old. macOS 10.14 will get released in a few months. If I can use `@available`, the beauty of it is that it will compile whether the API is available or not and also it injects runtime checks to make sure it's available at runtime too.

sacredbanana commented:
Oh and don't worry about Macs without a TouchBar. The system quietly ignores TouchBar calls if it doesn't have one

alkama commented:
I'm trying to wrap my head around what happens to that code https://github.com/Gargaj/Bonzomatic/pull/67/files#diff-e4c58fe627df60193a38f36b854be8c7R153 once compiled while running on older system versions ;)

sacredbanana commented:
`@available` will take care of that due to the runtime checks it injects

sacredbanana commented:
Also would very ancient Macs really be used as the shader coding computers in a shader showdown?

alkama commented:
Just like on PC, some mac users refuse to do major upgrades that work fine for them. And the computer itself is still valid platform indeed (my ex 2012 mac was a working Bonzomatic laptop with its gt 650m).
Anyway, when possible and reasonable we should do our best to not force deprecate platforms. Especially with features that are not "core".
But that's only my 2 cents as a user.

sacredbanana commented:
If you guys really don't want me to use `@available` I'll try to work out a less elegant solution

alkama commented:
Migrating the CI to XCode9 is not a bad idea indeed. And I dont mind '@available'. We will still need preprocessor removal of the whole touchbar for XCode <9 users (the same that refuse to upgrade, since XCode9 is only available for later platforms).

sacredbanana commented:
I'm not too familiar with cmake. How would I add a touchbar enabled option which if true does a `#define TOUCHBAR_ENABLED`? I can then `#ifdef TOUCHBAR_ENABLED` to the touchbar code and keep the `@available` since we can make this feature require Xcode 9 to build (but still run fine on old macs during runtime)

alkama commented:
I just fixed it, but i don't know how to add commits to a PR from another repo.

alkama commented:
Cant find a way to add to your PR, so just pushed my changes for CMake support on a branch on my fork. Feel free to incorporate it to this PR and discard mine.

sacredbanana commented:
Thank you very much. I have pulled in your change into my branch and added a note informing the user that Xcode 9 is required if TouchBar support is enabled. I also added the runtime `@available` check but as expected this causes the Travis CI Mac build to fail because it needs Xcode 9 to be able to understand the `@available` command in Objective-C

Gargaj commented:
How about adding https://docs.travis-ci.com/user/reference/osx/#Xcode-version ?
